docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,228 @@
+```luau
+type function neverending(t)
+    return neverending(t) -- note: parentheses are used here because the runtime value of types is being passed in, rather than the static annotation of types
+end
+```
+As a result, the current plan is to maintain a tightly scoped implementation, focusing on the simplest version of type functions to minimize the risk of developers accidentally starving themselves of features provided by the analysis, and gradually enhance capabilities through successive iterations.
+
+We have considered implementing a user-configurable execution limit based on time or instruction count for reducing type functions where if a type function exceeds the limit, it fails to reduce. However, these methods are not consistently reliable for terminating type functions that take too long to execute. Time-based timeouts are dependent on CPU performance; programs that type check on fast CPUs may not type check on slower CPUs. Similarly, instruction-based timeouts are dependent on the compiler optimizations; programs that type check on one compiler version may not type check on versions without the same optimizations. We will be experimenting with various forms of limiting the execution of type functions and reserve the right to change how termination of type functions is managed.
+
+<details><summary>List of illegal constructs in type functions</summary>
+
+* `while` / `repeat` loops
+* invoking other type functions / regular functions / lambdas
+    * we will not (and probably never) allow type functions to call regular functions for the sake of maintaining discrete stages between runtime and analysis
+* referring to locals / globals in the outer scope
+* global functions: `getfenv`, `setfenv`, `pcall`, `xpcall`, `require`
+* libraries: `coroutine`, `debug`, `string.gsub`
+
+Note: we are aware that for loops can cause infinite runtime. For the time being, we will not be handling this case. In the event that a developer accidentally creates an infinitely long type function, autocomplete will timeout in their editor environments and running luau-analyze will not complete. They will need to fix the type function and restart their environment / analysis.
+
+</details>
+
+To give warnings, developers can use `print()` with custom warning messages. To fail reductions, developers can use `error()` with custom error messages. If nothing is returned by the type function, it will fail to reduce with the default message: "Failed to reduce \<Name\> type function with no return value". If the return value of the type function is not an instance of typelib, it will fail to reduce with the message: "Failed to reduce \<Name\> type function with non-typelib return value".
```

Comment:
Can you explain the concept of a 'warning' better?
Today, there is no system for warning messages in the type inference engine (old or new).
If a new system is added, please describe the design of the warning system.

Reply:
I think right now it'd probably end up working exactly the same as everything does: just producing an error message either way, but there's a lot of reasons we probably ought to introduce a technical distinction between warnings and errors in the future. That's probably just another RFC altogether though, since it's likely going to include designing things like e.g. disabling warnings both locally and globally, forcing them to be errors, etc.

Reply:
Yep, I think that's a good call!